### PR TITLE
[SPARK-11654][SQL] add reduce to GroupedDataset

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala [hunk 1]
```diff
@@ -127,15 +125,26 @@ class GroupedDataset[K, T] private[sql](
...
     map((key, data) => f.call(key, data.asJava))(encoder)
   }
 
+  /**
+   * Reduces the elements of each group of data using the specified binary function.
+   * The given function must be commutative and associative or the result may be non-deterministic.
+   */
+  def reduce(f: (T, T) => T): Dataset[(K, T)] = {
+    val func = (key: K, it: Iterator[T]) => Iterator(key -> it.reduce(f))
+    flatMap(func)(ExpressionEncoder.tuple(kEnc, tEnc))
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: `ExpressionEncoder.tuple` doesn't work correctly now, I'll fix it in another PR.

**Comment by rxin** [comment]
```
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 4 in stage 54.0 failed 1 times, most recent failure: Lost task 4.0 in stage 54.0 (TID 130, localhost): java.lang.ClassCastException: scala.Tuple2 cannot be cast to java.lang.Integer
```

File: sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala (hunk 1 above)
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/GroupedDataset.scala]: I'm going to pick this up.

**Comment by cloud-fan** [comment]
@rxin the test is expected to fail because of a bug of `ExpressionEncoder.tuple`, see code comments above. @marmbrus is working on it.

**Pull request closed without merging**